flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/state/rocksdb/iterator/RocksMultiStateKeysIterator.java

```diff
@@ -0,0 +1,180 @@
+        try {
+            while (nextKey == null && hasDataToProcess()) {
+                pullKeysFromIterators();
+                K smallestIteratorKey = calculateSmallestKeyFromLocalData();
+                if (smallestIteratorKey != null) {
+                    previousKey = smallestIteratorKey;
+                    nextKey = smallestIteratorKey;
+                }
+            }
+        } catch (Exception e) {
+            throw new FlinkRuntimeException(
+                    "Failed to access states [" + String.join(",", states) + "]", e);
+        }
+        return nextKey != null;
+    }
+
+    private boolean hasDataToProcess() {
+        boolean result = iterators.stream().anyMatch(RocksIteratorWrapper::isValid);
+        if (!result) {
+            for (int i = 0; i < iterators.size(); ++i) {
+                if (iteratorKeys[i] != null) {
+                    result = true;
+                    break;
+                }
+            }
+        }
+        return result;
+    }
+
+    private void pullKeysFromIterators() {
+        for (int i = 0; i < iterators.size(); ++i) {
+            RocksIteratorWrapper iterator = iterators.get(i);
+            if (iteratorKeys[i] == null && iterator.isValid()) {
+                iteratorKeys[i] = iterator.key();
+                iterator.next();
+            }
+        }
+    }
+
+    @Nullable
+    private K calculateSmallestKeyFromLocalData() throws IOException {
+        int smallestIteratorKeyIndex = -1;
+        byte[] smallestIteratorKey = null;
+        for (int i = 0; i < iteratorKeys.length; ++i) {
+            byte[] iteratorKey = iteratorKeys[i];
+            if (iteratorKey != null
+                    && (smallestIteratorKey == null
+                            || Arrays.compare(iteratorKey, smallestIteratorKey) < 0)) {
+                smallestIteratorKeyIndex = i;
+                smallestIteratorKey = iteratorKey;
+            }
+        }
+
+        if (smallestIteratorKey != null) {
+            for (int i = 0; i < iteratorKeys.length; ++i) {
+                if (Arrays.equals(iteratorKeys[i], smallestIteratorKey)) {
+                    iteratorKeys[i] = null;
+                }
+            }
+
+            byteArrayDataInputView.setBuffer(
+                    smallestIteratorKey,
+                    keyGroupPrefixBytes,
+                    smallestIteratorKey.length - keyGroupPrefixBytes);
+            final K smallestIteratorKeyValue =
+                    CompositeKeySerializationUtils.readKey(
+                            keySerializer,
+                            byteArrayDataInputView,
+                            ambiguousKeyPossibles.get(smallestIteratorKeyIndex));
+            if (isMatchingNameSpace(
+                            smallestIteratorKey,
+                            byteArrayDataInputView.getPosition(),
+                            namespaceBytes)
+                    && !Objects.equals(previousKey, smallestIteratorKeyValue)) {
```

Comment:
I'm not sure is there any chance that  `Objects.equals(previousKey, smallestIteratorKeyValue)` is true?

Reply:
It took me some time why I've put it here but now I remember. It comes from this logic:
https://github.com/apache/flink/blob/cd0cfe42c15042a3c27a0c341416aa32c0d594b6/flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/state/rocksdb/iterator/RocksStateKeysIterator.java#L59-L77
It handles the case when we get the same key from an iterator. Honestly I've no idea whether and/or how this is possible but not wanted to miss this guard. If we could eliminate this then I think that could be quite a boost...

Reply:
I've the not confirmed feeling that `Objects.equals` can be removed here together with the one in [RocksStateKeysIterator.java](https://github.com/apache/flink/blob/cd0cfe42c15042a3c27a0c341416aa32c0d594b6/flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/state/rocksdb/iterator/RocksStateKeysIterator.java#L59-L77) but I would open a new jira to separate the concerns. WDYT?

Reply:
Agreed.